rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake

```diff
@@ -112,53 +108,72 @@ target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   "$<INSTALL_INTERFACE:include>"
 )
 foreach(_pkg_name ${rosidl_generate_interfaces_DEPENDENCY_PACKAGE_NAMES})
+  # Depend on targets generated by this generator in dependency packages
   target_link_libraries(
     ${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE
-    ${${_pkg_name}_TARGETS})
+    ${${_pkg_name}_TARGETS${_target_suffix}})
 endforeach()
+
 target_link_libraries(
   ${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE
   rosidl_runtime_cpp::rosidl_runtime_cpp)
 
+# Make the INTERFACE library created above depend on the generated headers.
+# That way if a package that generates interfaces also has a target depending
+# on those generated interfaces, that target will wait for this generator's
+# headers to be generated.
+add_dependencies(
+  ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+  ${rosidl_generate_interfaces_TARGET}__cpp)
+
+# Make top level generation target depend on this generated library
+add_dependencies(
+  ${rosidl_generate_interfaces_TARGET}
+  ${rosidl_generate_interfaces_TARGET}${_target_suffix})
+
 if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
-  if(NOT _generated_headers STREQUAL "")
-    install(
-      DIRECTORY ${_output_path}/
-      DESTINATION "include/${PROJECT_NAME}"
-      PATTERN "*.hpp"
-    )
-  endif()
-  ament_export_include_directories(include)
+  install(
+    DIRECTORY ${_output_path}/
+    DESTINATION "include/${PROJECT_NAME}"
+    PATTERN "*.hpp"
+  )
+
+  # Export old-style CMake variables
+  ament_export_include_directories("include")
+
+  # Export modern CMake targets
+  ament_export_targets(export_${rosidl_generate_interfaces_TARGET}${_target_suffix})
+  rosidl_export_typesupport_targets(${_target_suffix}
+    ${rosidl_generate_interfaces_TARGET}${_target_suffix})
 
   install(
     TARGETS ${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    EXPORT ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+    EXPORT export_${rosidl_generate_interfaces_TARGET}${_target_suffix}
```

Comment:
@sloretz nit: why add an `export_` suffix here but not for `rosidl_generator_c`?

Reply:
I'm inconsistent. Added `export_` in 368d022